src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -45,19 +34,29 @@ public sealed class ProcessInvoker : AgentService, IProcessInvoker
         public event EventHandler<DataReceivedEventArgs> OutputDataReceived;
         public event EventHandler<DataReceivedEventArgs> ErrorDataReceived;
 
-        public async Task<int> WaitForExit(CancellationToken cancellationToken)
+        private async Task<int> WaitForExit(CancellationToken cancellationToken)
         {
             // Wait for the cancellation token to be set or the process to exit.
-            while (!cancellationToken.IsCancellationRequested && !_proc.HasExited)
+            try
             {
-                await _processExitedSignal.WaitAsync(TraceInterval, cancellationToken);
-                if (!_proc.HasExited)
+                while (!cancellationToken.IsCancellationRequested && !_proc.HasExited)
                 {
-                    Trace.Verbose($"Waiting on process {_proc.Id} ({_stopWatch.Elapsed} elapsed)");
+                    await _processExitedSignal.WaitAsync(TraceInterval, cancellationToken);
+                    if (!_proc.HasExited)
+                    {
+                        Trace.Verbose($"Waiting on process {_proc.Id} ({_stopWatch.Elapsed} elapsed)");
+                    }
                 }
-            }
 
-            cancellationToken.ThrowIfCancellationRequested();
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+            finally
+            {
+                if (cancellationToken.IsCancellationRequested && !_proc.HasExited)
+                {
+                    _proc.Kill();
```

Comment:
Checking HasExited and then calling Kill is not atomic. Kill throws an InvalidOperationException if it's called and the process has already exited.